File: App/Core/Backup.swift
import Foundation
import CoreTransferable
import UniformTypeIdentifiers

// FFBackup — one portable file for everything she's made here: cycle history,
// settings, and the whole petal garden. Plain JSON with the two stores' own
// persisted blobs inside, so the same forward-compatible decoding that guards
// UserDefaults guards backups too. No cloud, no account — a file she owns.
struct FFBackupFile: Codable {
    var version = 1
    var exportedAt = Date()
    var cycle: Data      // CycleStore snapshot (logs + settings)
    var rewards: Data    // RewardsStore blob (points + everything owned)
}

enum FFBackup {
    /// Write a backup to a temp file for the share sheet. Nil on any failure.
    static func makeFile(store: CycleStore, rewards: RewardsStore) -> URL? {
        guard let cycle = store.backupData(),
              let garden = rewards.backupData() else { return nil }
        let enc = JSONEncoder()
        enc.dateEncodingStrategy = .iso8601
        guard let data = try? enc.encode(FFBackupFile(cycle: cycle, rewards: garden))
        else { return nil }
        let df = DateFormatter()
        df.dateFormat = "yyyy-MM-dd"
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("Uncorked-backup-\(df.string(from: Date())).json")
        do {
            try data.write(to: url, options: .atomic)
            return url
        } catch {
            return nil
        }
    }

    /// A share item that builds the backup file AT SHARE TIME, so what she
    /// saves is always current — never a snapshot frozen when the button first
    /// rendered (the tuning steppers and Restore live on the same screen).
    struct ShareItem: Transferable {
        let store: CycleStore
        let rewards: RewardsStore

        static var transferRepresentation: some TransferRepresentation {
            FileRepresentation(exportedContentType: .json) { item in
                guard let url = FFBackup.makeFile(store: item.store, rewards: item.rewards)
                else { throw CocoaError(.fileWriteUnknown) }
                return SentTransferredFile(url)
            }
        }
    }

    /// Restore both stores from a backup file's contents — all-or-nothing:
    /// BOTH payloads are validated before EITHER store is touched, so a
    /// corrupt or half-truncated file can never leave the two sides
    /// inconsistent (the destructive confirm dialog promises exactly that).
    @discardableResult
    static func restore(data: Data, store: CycleStore, rewards: RewardsStore) -> Bool {
        let dec = JSONDecoder()
        dec.dateDecodingStrategy = .iso8601
        guard let file = try? dec.decode(FFBackupFile.self, from: data),
              CycleStore.isValidBackup(file.cycle),
              RewardsStore.isValidBackup(file.rewards)
        else { return false }
        return store.restore(from: file.cycle) && rewards.restore(from: file.rewards)
    }
}

File: web/core/backup.js
// FFBackup on the web — a port of App/Core/Backup.swift. One portable file for
// everything she's made here: cycle history, settings, and the whole petal
// garden. Plain JSON with the two stores' own persisted blobs inside.
//
// CROSS-PLATFORM WIRE FORMAT — must match Swift's JSONEncoder exactly, so a
// backup written on Android restores on iOS and vice versa:
//   { "version": 1, "exportedAt": "<ISO-8601>", "cycle": "<base64>", "rewards": "<base64>" }
// Swift encodes the inner `Data` fields as BASE64 STRINGS (JSONEncoder's Data
// default) and dates as .iso8601. We do the same via btoa/atob over UTF-8.
//
// All-or-nothing restore: BOTH payloads are validated before EITHER store is
// touched, so a corrupt or half-truncated file can never leave the two sides
// inconsistent.

import { CycleStore } from "./store.js";
import { RewardsStore } from "./rewards.js";

export const BACKUP_VERSION = 1;

// UTF-8-safe base64 (btoa alone chokes on non-Latin1 — notes can hold anything).
const b64encode = (text) => {
  const bytes = new TextEncoder().encode(text);
  let bin = "";
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin);
};
const b64decode = (b64) => {
  const bin = atob(b64);
  const bytes = Uint8Array.from(bin, (c) => c.charCodeAt(0));
  return new TextDecoder().decode(bytes);
};

/// The backup file's name for today — same shape Swift's makeFile builds.
export function backupFilename(now = new Date()) {
  const y = now.getFullYear();
  const m = String(now.getMonth() + 1).padStart(2, "0");
  const d = String(now.getDate()).padStart(2, "0");
  return `Uncorked-backup-${y}-${m}-${d}.json`;
}

/// Serialize both stores into the portable backup JSON. Built AT CALL TIME so
/// what she saves is always current (Backup.swift's ShareItem promise).
export function makeBackup(store, rewards) {
  return JSON.stringify({
    version: BACKUP_VERSION,
    exportedAt: new Date().toISOString(),
    cycle: b64encode(store.backupData()),
    rewards: b64encode(rewards.backupData()),
  });
}

/// Restore both stores from a backup file's text — all-or-nothing. Returns
/// false (and changes nothing) on any invalid/corrupt input.
export function restoreBackup(text, store, rewards) {
  let file;
  try { file = JSON.parse(text); } catch { return false; }
  if (!file || typeof file.cycle !== "string" || typeof file.rewards !== "string") return false;
  let cycleText, rewardsText;
  try {
    cycleText = b64decode(file.cycle);
    rewardsText = b64decode(file.rewards);
  } catch { return false; }
  if (!CycleStore.isValidBackup(cycleText) || !RewardsStore.isValidBackup(rewardsText)) return false;
  return store.restore(cycleText) && rewards.restore(rewardsText);
}
